server/backends/github/github.go

```diff
@@ -324,13 +297,57 @@ func (c *OAuthHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, appRedirectURL, http.StatusTemporaryRedirect)
 }
 
-// requestAccessToken exchanges an OAuth code for an access token and links the
+// HandleLinkRepo handles the case where the user has never connected their
+// GitHub repo to BuildBuddy. This will take them to an oauth flow in GitHub.
+//
+// This flow will create a connection between a GitHub user and BuildBuddy user.
+// It will generate a GitHub token that a BuildBuddy user can use for GitHub operations.
+// It will not grant specific repo permissions. That must be separately granted
+// via the `handleInstallApp` flow, though this oauth step is a prerequisite.
+//
+// GitHub will redirect back to the callback URL set in the app settings (something like
+// auth/github/app/link or auth/github/read_only_app/link) with the "code" query param set.
+// This subsequent request should be handled with handleLinkRepoCallback
+func (c *OAuthHandler) HandleLinkRepo(w http.ResponseWriter, r *http.Request, redirectPath string) {
+	state := fmt.Sprintf("%d", random.RandUint64())
+	setCookies(w, r, state)
+
+	authURL := fmt.Sprintf(
+		"https://%s/login/oauth/authorize?client_id=%s&state=%s&redirect_uri=%s&scope=%s",
+		GithubHost(),
+		c.ClientID,
+		state,
+		url.QueryEscape(build_buddy_url.WithPath(redirectPath).String()),
+		"repo")
+	http.Redirect(w, r, authURL, http.StatusTemporaryRedirect)
+}
+
+func setCookies(w http.ResponseWriter, r *http.Request, state string) {
+	userID := r.FormValue("user_id")
+	groupID := r.FormValue("group_id")
+	redirectURL := r.FormValue("redirect_url")
+	if err := build_buddy_url.ValidateRedirect(redirectURL); err != nil {
+		redirectWithError(w, r, err)
+		return
+	}
```

Comment:
all callers of `setCookies` should also short-circuit when this happens

Reply:
Oh good catch - fixed